experiment/selinux/src/selinux.rs

```diff
@@ -44,268 +122,292 @@ impl SELinux {
+                        continue;
+                    }
+                    let key = fields[0].trim().to_string();
+                    let value = fields[1].trim().to_string();
+                    self.configs.insert(key, value);
                 }
             }
+            self.read_config_init_done.store(true, Ordering::SeqCst);
         }
-        Ok(())
+        self.configs
+            .get(target_key)
+            .cloned()
+            .filter(|s| !s.is_empty())
+            .ok_or(SELinuxError::GetConfigKey(format!(
+                "can't find the target label in the config file: {}",
+                target_key
+            )))
     }
 
-    // function similar with lSetFileLabel in go-selinux repo.
-    // lset_file_label sets the SELinux label for this path, not following symlinks,
-    // or returns an error.
-    pub fn lset_file_label(fpath: &Path, label: &str) -> Result<(), SELinuxError> {
-        if !fpath.exists() {
-            return Err(SELinuxError::LSetFileLabel(ERR_EMPTY_PATH.to_string()));
+    // get_enabled returns whether SELinux is enabled or not.
+    pub fn get_enabled(&mut self) -> bool {
+        match Self::get_selinux_mountpoint(self) {
+            // If there is no SELinux mountpoint, SELinux is not enabled.
+            None => false,
+            Some(_) => match Self::current_label(self) {
+                Ok(con) => {
+                    // Check whether label is "kernel" or not.
+                    if con.user != "kernel" {
+                        return true;
+                    }
+                    false
+                }
+                Err(_) => false,
+            },
         }
+    }
 
+    // verify_selinux_fs_mount verifies if the specified mount point is
+    // properly mounted as a writable SELinux filesystem.
+    fn verify_selinux_fs_mount<P: AsRef<Path>>(mnt: P) -> bool {
+        let mnt = mnt.as_ref();
         loop {
-            match fpath.lset_xattr(XATTR_NAME_SELINUX, label.as_bytes()) {
-                Ok(_) => break,
-                // When a system call is interrupted by a signal, it needs to be retried.
-                Err(XattrError::EINTR(_)) => continue,
-                Err(e) => {
-                    return Err(SELinuxError::LSetFileLabel(format!(
-                        "lset_xattr failed: {}",
-                        e
-                    )));
+            match statfs::statfs(mnt) {
+                Ok(stat) => {
+                    // verify if the file is readonly or not
+                    if !stat.flags().contains(statvfs::FsFlags::ST_RDONLY) {
+                        return false;
+                    }
+                    // verify if the file is SELinux filesystem
+                    return stat.filesystem_type() == statfs::SELINUX_MAGIC;
                 }
+                // check again if there is an issue while calling statfs
+                Err(Errno::EAGAIN) | Err(Errno::EINTR) => continue,
```

Comment:
Why do we repeat this case? I'm afraid of the infinity loop.

Reply:
@utam0k 
I referred to go-selinux's implementation.
I understood that we need infinity loop in case of temporary issues like interruptions.
Do you think I don't need the loop here?

https://github.com/opencontainers/selinux/blob/bb1ec25c889f29e35dffec33346af3c7d434dbc1/go-selinux/selinux_linux.go#L135

Reply:
Okay, I understood. Let's keep it. Thanks for your detailed explanation.